File: backend/app/schemas/printing_purchase_orders.py
# Request/response bodies for the printing orders module's endpoints.
#
# Shaped like schemas/purchase_orders.py on purpose — same parallel-array
# line items, same GST arrangement — with one difference that runs through
# everything: a line item is a `description` the vendor wrote, not a
# product_id of ours. Nothing here can name a product, which is what keeps a
# printing order from ever reaching the inventory code.
from datetime import datetime

from pydantic import BaseModel, model_validator

from app.services.gst import TaxKind


def _check_line_item_arrays(
    descriptions: list[str],
    hsn_codes: list[str] | None,
    quantities: list[int],
    rates: list[float],
    gst_percs: list[float],
) -> None:
    lengths = {len(descriptions), len(quantities), len(rates), len(gst_percs)}
    if hsn_codes is not None:
        lengths.add(len(hsn_codes))
    if len(lengths) > 1:
        raise ValueError(
            "descriptions, hsn_codes, quantities, rates and gst_percs must have the same number of entries"
        )
    if len(descriptions) == 0:
        raise ValueError("at least one line item is required")
    if any(not description.strip() for description in descriptions):
        # The description IS the line item here — an empty one would record a
        # charge against nothing at all, where a material order at least
        # still points at a product.
        raise ValueError("every line item needs a description")


class CreateNewPrintingPurchaseOrderRequest(BaseModel):
    purchase_order_no: str
    vendor_id: int
    date: datetime
    # Parallel arrays, one entry per line item. `descriptions` is the service
    # as the vendor billed it and is required; `hsn_codes` is the SAC/HSN
    # they printed against it and is optional, since plenty of printing bills
    # don't carry one.
    descriptions: list[str]
    hsn_codes: list[str] | None = None
    quantities: list[int]
    rates: list[float]
    # One GST rate per line item — the rate that actually applies to each
    # line, and what the totals are computed from. Per line because a
    # printing bill can mix rates just as a material one can.
    gst_percs: list[float]
    # Which heads the rates fall under. Sent explicitly because a mixed-rate
    # order has no single header percentage to infer it from; the endpoint
    # falls back to the two parties' states when it's omitted.
    tax_kind: TaxKind | None = None
    description: str
    # The vendor's own invoice number, set only when this order came from an
    # uploaded PDF. Carried onto the printing purchase invoice this order
    # raises, where it's what makes a re-upload of the same document a
    # conflict.
    vendor_invoice_no: str | None = None

    @model_validator(mode="after")
    def _check_line_items_match(self) -> "CreateNewPrintingPurchaseOrderRequest":
        _check_line_item_arrays(self.descriptions, self.hsn_codes, self.quantities, self.rates, self.gst_percs)
        return self


class CreateNewPrintingPurchaseOrderResponse(BaseModel):
    message: str
    # Every printing purchase order raises its printing purchase invoice as
    # it's created (see services/printing_purchase_invoices.py). Its id comes
    # back so the caller can attach the vendor's PDF to it in a follow-up
    # request — the only reason this response carries an id at all.
    printing_purchase_invoice_id: int


class PrintingPurchaseOrderDetailItem(BaseModel):
    id: int
    purchase_order_no: str
    vendor_id: int
    date: datetime
    descriptions: list[str]
    hsn_codes: list[str]
    quantities: list[int]
    rates: list[float]
    # One per line item, parallel to the arrays above. The edit form fills
    # its per-line GST column from these; sgst_perc/cgst_perc/igst_perc below
    # are the derived single-rate summary and are all None when these differ.
    gst_percs: list[float]
    total_amount_before_tax: float
    tax_kind: TaxKind
    sgst_perc: float | None = None
    cgst_perc: float | None = None
    igst_perc: float | None = None
    total_amount_after_tax: float
    description: str


class UpdatePrintingPurchaseOrderDetailsRequest(BaseModel):
    id: int
    purchase_order_no: str
    vendor_id: int
    date: datetime
    descriptions: list[str]
    hsn_codes: list[str] | None = None
    quantities: list[int]
    rates: list[float]
    gst_percs: list[float]
    tax_kind: TaxKind | None = None
    description: str

    @model_validator(mode="after")
    def _check_line_items_match(self) -> "UpdatePrintingPurchaseOrderDetailsRequest":
        _check_line_item_arrays(self.descriptions, self.hsn_codes, self.quantities, self.rates, self.gst_percs)
        return self


class UpdatePrintingPurchaseOrderDetailsResponse(BaseModel):
    message: str


class ParsedPrintingInvoiceLineItem(BaseModel):
    # Everything the invoice printed for this line, taken as printed. There
    # is no product_id and no unresolved_reason — unlike a material line,
    # this one has nothing to resolve against, which is the whole reason
    # printing has its own intake.
    description: str
    hsn_code: str = ""
    quantity: int
    rate: float
    gst_perc: float


class ParsePrintingPurchaseInvoicePdfResponse(BaseModel):
    # Everything read off an uploaded printing invoice, in the shape the
    # printing purchase order form fills itself from. Nothing is written to
    # the database by the parse call itself — the admin reviews these values
    # and submits them through create_new_printing_purchase_order.
    vendor_id: int
    vendor_name: str
    vendor_gstin: str
    vendor_invoice_no: str
    date: datetime
    line_items: list[ParsedPrintingInvoiceLineItem]
    # The heads this purchase falls under, from our state vs the vendor's.
    # Always present, unlike the percentages below, which are all None when
    # the invoice taxes its lines at different rates.
    tax_kind: TaxKind
    sgst_perc: float | None = None
    cgst_perc: float | None = None
    igst_perc: float | None = None
    total_amount_before_tax: float
    total_amount_after_tax: float
    # The invoice's own printed grand total and whether it disagrees with the
    # total derived from the line items. A mismatch is shown to the admin as
    # a warning, not a rejection.
    printed_total: float | None = None
    total_mismatch: bool = False
    # "text" if the deterministic parser read this invoice, "claude" if it
    # took the fallback — shown on the review screen so the admin knows how
    # much to trust the values.
    source: str

File: frontend/src/lib/printing-purchase-orders.ts
// ---------------------------------------------------------------------------
// Printing purchase order data for the /admin/orders "Purchase orders" tab
// ---------------------------------------------------------------------------
// The printing half of lib/purchase-orders.ts, against
// backend/app/api/routes/printing_orders.py and its own
// #printing_purchase_orders collection.
//
// The one structural difference runs through everything here: a line item is
// a DESCRIPTION the vendor wrote, not a product of ours. A printing vendor
// bills a service ("Customized Print Service", "Sticker Printing A3 UV")
// that matches nothing in our catalogue and isn't supposed to, so there are
// no productIds to resolve, nothing to look up against lib/products.ts, and
// nothing that moves stock. That is also why these are separate collections
// rather than a flag on the material ones — see the backend model's comments.
//
// Everything about GST is identical to the material side: the RATE belongs
// to the line item (a printing bill can mix rates as freely as a material
// one), and the order only decides which HEADS carry it, from the two
// parties' states.
import { apiFetch } from "@/lib/api";
import type { TaxKind } from "@/lib/purchase-orders";

export type PrintingPurchaseOrder = {
  id: number;
  purchaseOrderNo: string;
  vendorId: number;
  date: string;
  // Parallel arrays, one entry per line item. `descriptions` is the service
  // as the vendor billed it; `hsnCodes` is the SAC/HSN printed against it,
  // "" where the bill carried none.
  descriptions: string[];
  hsnCodes: string[];
  quantities: number[];
  rates: number[];
  // One GST rate per line item — the rate each line is actually taxed at,
  // and what the totals are built from.
  gstPercs: number[];
  totalAmountBeforeTax: number;
  // Which heads the rates fall under. Never null here, unlike on the
  // material side: this collection has no rows predating the field.
  taxKind: TaxKind;
  // The order's single rate under its heads — a derived summary, all null
  // when the line items are taxed at different rates.
  sgstPerc: number | null;
  cgstPerc: number | null;
  igstPerc: number | null;
  totalAmountAfterTax: number;
  description: string;
};

// Shape returned by the backend's PrintingPurchaseOrderDetailItem schema.
type PrintingPurchaseOrderDetailItem = {
  id: number;
  purchase_order_no: string;
  vendor_id: number;
  date: string;
  descriptions: string[];
  hsn_codes: string[];
  quantities: number[];
  rates: number[];
  gst_percs: number[];
  total_amount_before_tax: number;
  tax_kind: TaxKind;
  sgst_perc: number | null;
  cgst_perc: number | null;
  igst_perc: number | null;
  total_amount_after_tax: number;
  description: string;
};

export async function fetchPrintingPurchaseOrders(): Promise<PrintingPurchaseOrder[]> {
  const response = await apiFetch("/admin/get_printing_purchase_order_details");
  if (!response.ok) {
    throw new Error("Failed to load printing purchase orders");
  }

  const items: PrintingPurchaseOrderDetailItem[] = await response.json();
  return items.map((item) => ({
    id: item.id,
    purchaseOrderNo: item.purchase_order_no,
    vendorId: item.vendor_id,
    date: item.date,
    descriptions: item.descriptions,
    hsnCodes: item.hsn_codes,
    quantities: item.quantities,
    rates: item.rates,
    gstPercs: item.gst_percs,
    totalAmountBeforeTax: item.total_amount_before_tax,
    taxKind: item.tax_kind,
    sgstPerc: item.sgst_perc,
    cgstPerc: item.cgst_perc,
    igstPerc: item.igst_perc,
    totalAmountAfterTax: item.total_amount_after_tax,
    description: item.description,
  }));
}

// ---------------------------------------------------------------------------
// Reading a printing purchase order off the vendor's own invoice PDF
// ---------------------------------------------------------------------------
// POST /admin/parse_printing_purchase_invoice_pdf reads an uploaded printing
// invoice and returns the values an order would be built from. It writes
// nothing: the admin reviews them in the ordinary printing purchase order
// form and saves through create_new_printing_purchase_order.
//
// Unlike the material parse, no line can come back "unresolved" — there is
// nothing to resolve a description against, so every line arrives complete
// and editable. What it CAN refuse is an invoice whose vendor isn't on file,
// whose vendor is a material vendor (the bill belongs on the other tab), one
// already recorded, or a PDF that couldn't be read in full.

export type ParsedPrintingInvoiceLineItem = {
  description: string;
  hsnCode: string;
  quantity: number;
  rate: number;
  gstPerc: number;
};

export type ParsedPrintingPurchaseInvoice = {
  vendorId: number;
  vendorName: string;
  vendorGstin: string;
  vendorInvoiceNo: string;
  date: string;
  lineItems: ParsedPrintingInvoiceLineItem[];
  // Which heads this purchase falls under, from our state vs the vendor's.
  // Always present, unlike the percentages below, which are all null when
  // the invoice taxes its lines at different rates.
  taxKind: TaxKind;
  sgstPerc: number | null;
  cgstPerc: number | null;
  igstPerc: number | null;
  totalAmountBeforeTax: number;
  totalAmountAfterTax: number;
  // The invoice's own printed total, and whether it disagrees with what
  // these line items add up to. Shown as a warning to check rather than
  // treated as an error — printing bills carry freight and round-off lines
  // too.
  printedTotal: number | null;
  totalMismatch: boolean;
  // "text" if the deterministic parser read the PDF, "claude" if it fell
  // back to reading it with Claude.
  source: string;
};

// Shape returned by the backend's ParsePrintingPurchaseInvoicePdfResponse.
type ParsePrintingPurchaseInvoicePdfResponse = {
  vendor_id: number;
  vendor_name: string;
  vendor_gstin: string;
  vendor_invoice_no: string;
  date: string;
  line_items: {
    description: string;
    hsn_code: string;
    quantity: number;
    rate: number;
    gst_perc: number;
  }[];
  tax_kind: TaxKind;
  sgst_perc: number | null;
  cgst_perc: number | null;
  igst_perc: number | null;
  total_amount_before_tax: number;
  total_amount_after_tax: number;
  printed_total: number | null;
  total_mismatch: boolean;
  source: string;
};

export async function parsePrintingPurchaseInvoicePdf(
  file: File,
): Promise<ParsedPrintingPurchaseInvoice> {
  const formData = new FormData();
  formData.append("file", file);

  const response = await apiFetch("/admin/parse_printing_purchase_invoice_pdf", {
    method: "POST",
    body: formData,
  });

  if (!response.ok) {
    // Every rejection reason is written for the admin and names what to fix
    // ("no vendor with GSTIN ...", "X is a material vendor — record this
    // invoice under Purchase orders / Material instead"), so it's shown
    // as-is rather than flattened into a status-code message.
    const body = await response.json().catch(() => null);
    throw new Error(
      typeof body?.detail === "string" ? body.detail : "this invoice couldn't be read. Please try again.",
    );
  }

  const parsed: ParsePrintingPurchaseInvoicePdfResponse = await response.json();
  return {
    vendorId: parsed.vendor_id,
    vendorName: parsed.vendor_name,
    vendorGstin: parsed.vendor_gstin,
    vendorInvoiceNo: parsed.vendor_invoice_no,
    date: parsed.date,
    lineItems: parsed.line_items.map((item) => ({
      description: item.description,
      hsnCode: item.hsn_code,
      quantity: item.quantity,
      rate: item.rate,
      gstPerc: item.gst_perc,
    })),
    taxKind: parsed.tax_kind,
    sgstPerc: parsed.sgst_perc,
    cgstPerc: parsed.cgst_perc,
    igstPerc: parsed.igst_perc,
    totalAmountBeforeTax: parsed.total_amount_before_tax,
    totalAmountAfterTax: parsed.total_amount_after_tax,
    printedTotal: parsed.printed_total,
    totalMismatch: parsed.total_mismatch,
    source: parsed.source,
  };
}
